### PR TITLE
Api 806 r4 patient controller

### PR DESCRIPTION
**Pull request by eclendenning-va**

### CODE REVIEW

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+ * https://build.fhir.org/ig/HL7/US-Core-R4/StructureDefinition-us-core-patient.html for
+ * implementation details.
+ */
+@Slf4j
+@Validated
+@RestController
+@SuppressWarnings("WeakerAccess")
+@RequestMapping(
+    value = {"/r4/Patient"},
+    produces = {"application/json", "application/json+fhir", "application/fhir+json"})
```
Comment (jhulbert-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java:43]: Remove `json+fhir`

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@SuppressWarnings("WeakerAccess")
+@DataJpaTest
```
Comment (jblefkove-va) [inline data-query/src/test/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientControllerTest.java:32]: Let's include `@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)` on new `@DataJpaTest` classes

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+      @RequestParam("name") String name,
+      @RequestParam("birthdate") String[] birthdate,
+      @RequestParam(value = "page", defaultValue = "1") @Min(1) int page,
+      @CountParameter @Min(0) int count) {
+    PatientRepositoryV2.NameAndBirthdateSpecification spec =
+        PatientRepositoryV2.NameAndBirthdateSpecification.builder()
+            .name(name)
+            .dates(birthdate)
+            .build();
+    log.info("PatientV2: Looking for {} {}", sanitize(name), spec);
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java:217]: Referencing `PatientV2` directly seems like cruft?

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+      @RequestParam("birthdate") String[] birthdate,
+      @RequestParam("family") String family,
+      @RequestParam(value = "page", defaultValue = "1") @Min(1) int page,
+      @CountParameter @Min(0) int count) {
+    PatientRepositoryV2.BirthdateAndFamilySpecification spec =
+        PatientRepositoryV2.BirthdateAndFamilySpecification.builder()
+            .family(family)
+            .dates(birthdate)
+            .build();
+    log.info("PatientV2: Looking for {} {}", spec, sanitize(family));
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java:136]: ditto

File: data-query/spotbugs-excludes.xml [hunk 5]
```diff
@@ -64,6 +64,7 @@
       <Class name="gov.va.api.health.dataquery.service.controller.observation.ObservationRepository$PatientAndCategoryAndDateSpecification$PatientAndCategoryAndDateSpecificationBuilder"/>
       <Class name="gov.va.api.health.dataquery.service.controller.patient.PatientSearchRepository$NameAndBirthdateSpecification$NameAndBirthdateSpecificationBuilder"/>
       <Class name="gov.va.api.health.dataquery.service.controller.patient.PatientRepositoryV2$NameAndBirthdateSpecification$NameAndBirthdateSpecificationBuilder"/>
+      <Class name="gov.va.api.health.dataquery.service.controller.patient.PatientRepositoryV2$BirthdateAndFamilySpecification$BirthdateAndFamilySpecificationBuilder"/>
```
Comment (jblefkove-va) [inline data-query/spotbugs-excludes.xml:67]: alphabetical

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+      headers = {"raw=true"})
+  public String readRaw(@PathVariable("publicId") String publicId, HttpServletResponse response) {
+    PatientEntityV2 entityV2 = findById(publicId);
+    IncludesIcnMajig.addHeader(response, entityV2.icn());
+    return entityV2.payload();
+  }
+
+  /** Search by Birthdate+Family. */
+  @GetMapping(params = {"birthdate", "family"})
+  public Patient.Bundle searchByBirthdateandFamily(
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java:126]: birthdateAnd

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+  @GetMapping(params = {"_id"})
+  public Patient.Bundle searchById(
+      @RequestParam("_id") String id,
+      @RequestParam(value = "page", defaultValue = "1") @Min(1) int page,
+      @CountParameter @Min(0) int count) {
+    return searchByIdentifier(id, page, count);
+  }
+
+  /** Search by Identifier. */
+  @GetMapping(params = {"identifier"})
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java:175]: The spec indicates that identifier should be system+code: `GET [base]/Patient?identifier={[system]}|[code]`

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+  /** Search by Name. */
+  @GetMapping(params = {"name"})
+  public Patient.Bundle searchByName(
+      @RequestParam("name") String name,
+      @RequestParam(value = "page", defaultValue = "1") @Min(1) int page,
+      @CountParameter @Min(0) int count) {
+    return bundle(
+        Parameters.builder().add("name", name).add("page", page).add("_count", count).build(),
+        repository.findByFullName(name, page(page, count)));
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java:202]: Ultimately this is a string equality check, right? Spec says `string match of any part of name`

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java (hunk 7 above)
Comment (jhulbert-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientController.java:175]: Identifier I think so (the example shows it as well).
Gender is a bit confusing though, thoughts?
```
GET [base]/Patient?family=[string]&gender={[system]}|[code]

Example:
GET [base]/Patient?family=Shaw&gender=female
```

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+  @Test
+  public void searchByFamilyAndGenderWithCountZero() {
+    DatamartPatient dm = Datamart.create().patient("x");
+    testEntityManager.persistAndFlush(asPatientEntityV2(dm));
+    Patient.Bundle patient = controller().searchByFamilyAndGender("Wolff180", "male", 1, 0);
+    assertThat(patient.entry()).isEqualTo(Collections.emptyList());
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void searchByFamilyAndGenderWithNullCdw() {
```
Comment (jhulbert-va) [inline data-query/src/test/java/gov/va/api/health/dataquery/service/controller/patient/R4PatientControllerTest.java:128]: `WithNullGender`?

**Pull request merged**